src/main/java/org/candlepin/subscriptions/tally/admin/MergeHostsMigration.java

```diff
@@ -0,0 +1,240 @@
+                                and instance_id <> ''
+                                and hardware_type <> ''
+                                order by instance_id asc
+                                limit ?;
+      """;
+
+  private HostRepository hostRepository;
+  private final Counter counter;
+  private String orgId;
+
+  public MergeHostsMigration(
+      JdbcTemplate jdbcTemplate, MeterRegistry meterRegistry, HostRepository hostRepository) {
+    super(jdbcTemplate, meterRegistry);
+    counter = meterRegistry.counter("swatch_mergeHosts_migration");
+    this.hostRepository = hostRepository;
+  }
+
+  @Override
+  public SqlRowSet extract(String recordOffset, int batchSize) {
+    if (orgId == null) {
+      return jdbcTemplate.query(
+          INSTANCE_DUPLICATE_QUERY_NO_ORG,
+          new Object[] {recordOffset, recordOffset, batchSize},
+          new int[] {Types.VARCHAR, Types.VARCHAR, Types.NUMERIC},
+          SQL_ROW_SET_RESULT_SET_EXTRACTOR);
+    } else {
+      return jdbcTemplate.query(
+          INSTANCE_DUPLICATE_QUERY_WITH_ORG,
+          new Object[] {recordOffset, recordOffset, orgId, batchSize},
+          new int[] {Types.VARCHAR, Types.VARCHAR, Types.VARCHAR, Types.NUMERIC},
+          SQL_ROW_SET_RESULT_SET_EXTRACTOR);
+    }
+  }
+
+  public void setOrgId(String orgId) {
+    this.orgId = orgId;
+  }
+
+  @Transactional
+  @Override
+  public String transformAndLoad(SqlRowSet data) {
+    String lastInstanceId = null;
+    int hostCount = 0;
+
+    while (data.next()) {
+      List<Host> deleteList = new ArrayList<>();
+      String instanceId = data.getString("instance_id");
+      String organizationId = data.getString("org_id");
+      // get the hosts that share the id.
+      List<Host> hosts =
+          hostRepository
+              .findAllByOrgIdAndInstanceIdIn(organizationId, Set.of(instanceId))
+              .collect(Collectors.toCollection(ArrayList::new));
+      Collections.sort(hosts, Comparator.comparing(Host::getLastSeen).reversed());
+      Host primaryHost = hosts.get(0);
+      for (Host host : hosts) {
+        if (primaryHost.getId().equals(host.getId())) {
+          continue;
+        }
+        deleteList.add(host);
+        mergeDataFields(primaryHost, host);
+        mergeTallyBuckets(primaryHost, host);
+        mergeMeasurements(primaryHost, host);
+        mergeMonthlyTotals(primaryHost, host);
+      }
+
+      lastInstanceId = instanceId;
+      hostCount++;
+      try {
+        hostRepository.save(primaryHost);
+      } catch (Exception e) {
+        log.error("Unable to update host on host merge: [instanceId: {}]", e.getMessage());
+        throw e;
+      }
+      hostRepository.deleteAll(deleteList);
+    }
+
+    counter.increment(hostCount);
+    return lastInstanceId;
+  }
+
+  /**
+   * Merge the identifier values into the primary (latest seen wins)
+   *
+   * @param primaryHost
+   * @param host
+   */
+  private void mergeDataFields(Host primaryHost, Host host) {
```

Comment:
It might be helpful to adjust the merge methods so that instead of mutating the incoming Host objects, we return a new Host object. This way, we could have unit tests that assert directly on the output, making it easier to validate the behavior of each method in isolation. What do you think?